lib/src/rules/prefer_asserts_in_initializer_list.dart

```diff
@@ -68,30 +70,56 @@ class Visitor extends SimpleAstVisitor {
 }
 
 class _AssertVisitor extends RecursiveAstVisitor {
-  _AssertVisitor(this.classElement);
-  final ClassElement classElement;
+  _AssertVisitor(this.constructorElement);
+
+  final ConstructorElement constructorElement;
   bool needInstance = false;
 
+  ClassElement get classElement => constructorElement.enclosingElement;
+
+  bool _isInstanceNeeded(Expression node) {
+    if (node is MethodInvocation) {
+      return _isInstanceNeeded(node.target);
+    } else if (node is PropertyAccess) {
+      return _isInstanceNeeded(node.target);
```

Comment:
Given that `visitMethodInvocation` and `visitPropertyAccess` both cause all of their children to be visited, you shouldn't need to explicitly invoke `_isInstanceNeeded` on their children. In which case, the check inside the `node is SimpleIdentifier` branch can be moved into `visitSimpleIdentifier` and `_isInstanceNeeded` can be removed (as well as two of the visit methods).